### PR TITLE
OpenGL backend (stage 2.2.2)

### PR DESCRIPTION
**Pull request by archibate**
Related issue id = #492

### CODE REVIEW

**Comment by archibate** [comment]
@archibate requested a review from @yuanming-hu

**Comment by archibate** [comment]
Any idea?

File: taichi/codegen/codegen_opengl.cpp [hunk 1]
```diff
@@ -268,26 +380,25 @@ class KernelGen : public IRVisitor
 
   void visit(AllocaStmt *alloca) override
   {
-    emit("{} {}(0);",
+    emit("{} {};", // need = 0?
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_opengl.cpp]: Yes.

File: taichi/codegen/codegen_opengl.cpp [hunk 2]
```diff
@@ -216,29 +328,29 @@ class KernelGen : public IRVisitor
     const auto bin_name = bin->raw_name();
     if (bin->op_type == BinaryOpType::floordiv) {
       if (is_integral(bin->element_type())) {
-        emit("const {} {} = int(floor({} / {}));", dt_name, bin_name, lhs_name,
+        emit("{} {} = int(floor({} / {}));", dt_name, bin_name, lhs_name,
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_opengl.cpp]: Is there a good reason to remove those `const`'s?

**Comment by yuanming-hu** [comment]
Sorry about the delay. I was busy adding CUDA support for Windows.

File: taichi/codegen/codegen_opengl.cpp (hunk 2 above)
Comment (archibate) [reply taichi/codegen/codegen_opengl.cpp]: Since all `tmpxx` won't be assigned one more time. No need to specify `const` to some while some are not :)

Comment (yuanming-hu) [reply taichi/codegen/codegen_opengl.cpp]: > all tmpxx won't be assigned one more time

I would say that's exactly why we should mark them as `const` :-)

**Comment by archibate** [comment]
Did you see any potential bug in this stage? I always get a `material[i] = 2438xxxx7 (sounds INT_MAX)` in mpm128.py after all done...

**Comment by archibate** [comment]
Go for ast now..

**Comment by yuanming-hu** [comment]
> Did you see any potential bug in this stage? I always get a material[i] = 2438xxxx7 (sounds INT_MAX) in mpm128.py after all done...

Would it be possible to pinpoint which stage (SNode writer, kernel print, SNodeReader etc) caused the INT_MAX result?

**Comment by archibate** [comment]
> > Did you see any potential bug in this stage? I always get a material[i] = 2438xxxx7 (sounds INT_MAX) in mpm128.py after all done...
> 
> Would it be possible to pinpoint which stage (SNode writer, kernel print, SNodeReader etc) caused the INT_MAX result?

Can't, but I guess it might be related to range for or floordiv:
```
@ti.kernel
def reset():
  group_size = n_particles // 3
  for i in range(n_particles):
    x[i] = [ti.random() * 0.2 + 0.3 + 0.10 * (i // group_size), ti.random() * 0.2 + 0.05 + 0.32 * (i // group_size)]
    material[i] = i // group_size # problem is very likely to be raised from here
    # ...
```

**Comment by archibate** [comment]
(Remember to test non-GL build)

**Comment by yuanming-hu** [comment]
A seemingly stupid yet extremely effective way to debug issues like these: simply the program as much as you can while keeping the bug reproducible. Ultimately you will likely get a Python file within ~10 lines of code to trigger te bug. This will help you narrow down the range of possible sources of bugs. I've been using this technique very frequently to debug mysterious compiler bugs...

**Comment by archibate** [comment]
Go bed now, do we need a stage 3 or just done everything here (like test passing)?

**Comment by yuanming-hu** [comment]
Thansk for your hard work!

As long as we can pass the corresponding tests, then we are good :-) The initial release should also show some performance boost compared with the CPU version. We can leave further performance optimizations to the future versions.

**Comment by archibate** [comment]
Stage 3 merged! Check if test pass...

**Comment by archibate** [comment]
```
0:44(6): warning: identifier `func_c19_0__grad0' uses reserved `__' string
```
Have no idea why `__` is reserved...

**Comment by yuanming-hu** [comment]
> ```
> 0:44(6): warning: identifier `func_c19_0__grad0' uses reserved `__' string
> ```
> 
> 
> Have no idea why `__` is reserved...

The double underscore is weird anyway - let's remove the first/second underscore then.

**Comment by yuanming-hu** [comment]
https://github.com/taichi-dev/taichi/blob/master/python/taichi/lang/kernel.py#L221

**Comment by yuanming-hu** [comment]
The format template should be `{}_c{}_{}{}` instead of `{}_c{}_{}_{}` :-)

**Comment by archibate** [comment]
I think here is done if all tests passed.
Also want to compare `archibate:opengl` with `g2` to make sure that no commits are lost.

**Comment by archibate** [comment]
How to make spdlog submodule happy?

**Comment by archibate** [comment]
> A seemingly stupid yet extremely effective way to debug issues like these: simplify the program as much as you can while keeping the bug reproducible. Ultimately you will likely get a Python file within ~10 lines of code to trigger the bug. This will help you narrow down the range of possible sources of bugs. I've been using this technique very frequently to debug mysterious compiler bugs...

When I move `group_size` outside, the fault no longer occur. So it may be related to multi-offload support.

**Comment by archibate** [comment]
Hello?

**Comment by yuanming-hu** [comment]
Sorry I've been busy review a couple of SIGGRAPH papers (due at 6 pm today). I'll have more time after that.

**Comment by yuanming-hu** [comment]
> When I move `group_size` outside, the fault no longer (never) occur. So it may be related to multi-offload support or GlobalTmpStmt.

Back to Taichi after the SIGGRAPH review rush. Which `group_size` are you referring to here?

**Comment by yuanming-hu** [comment]
I see which `group_size` you are talking about. Yeah, the bug may be related to offloading.

File: taichi/perf.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <taichi/common/util.h>
+#include <taichi/system/timer.h>
+
+TI_NAMESPACE_BEGIN
+
+#define TI_MAX_PERF_ID 32
+
+static inline double TI_PERF(const char *left = nullptr, int perf_id = 0, int show_freq = 1)
```
Comment (yuanming-hu) [inline taichi/perf.h]: Is there a reason for a new profiler?

File: python/taichi/misc/test.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import taichi as ti
+
+def approx(*args, **kwargs):
+  from pytest import approx
+  if ti.cfg.arch == ti.opengl:
+    kwargs['rel'] = max(kwargs.get('rel', 1e6), 1e4)
```
Comment (yuanming-hu) [inline python/taichi/misc/test.py]: Did you mean `1e-6` and `1e-4` here?

File: examples/nbody_oscillator.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import taichi as ti
+
+ti.init(arch=ti.cuda) # Try to run on GPU, use ti.opengl if you don't have CUDA
+
+N = 8000
+pos = ti.var(dt=ti.f32, shape=(N, 2))
```
Comment (yuanming-hu) [inline examples/nbody_oscillator.py]: A better practice is `pos = ti.Vector(2, dt=ti.f32, shape=N)`. The same thing for `vel`. Then you can make use of vector arithmetic in the `advance` function.

Comment (archibate) [reply examples/nbody_oscillator.py]: Good idea! But when I have this example written, OpenGL hadn't support Matrix yet and I had to norm_sqr by hand :) Will change this trr.

**Comment by archibate** [comment]
Created #586. Btw, do you have some good suggestions for breaking down PR and again assemble them back into master? Had a hard time with `git cherry-pick` when changes requested.

**Comment by archibate** [comment]
I keep this open so it can be used as another (maybe last) stage after #586 merged.

**Comment by archibate** [comment]
Agree, will fix that bug in a distinct PR.

**Comment by yuanming-hu** [comment]
> Created #586. Btw, do you have some good suggestions for breaking down PR and again assemble them back into master? Had a hard time with `git cherry-pick` when changes requested.

Usually I just `git cherry-pick` a range commits to rebase them onto master. If there are conflicting changes then I have to do manual work to resolve them. This can be sometimes troublesome (That's exactly why we should keep PRs small so they get merged on time to avoid conflicts...)

(@k-ye Do you have a smarter way to do this?)

If it's too troublesome, maybe we can to plan A: I'll just squash and merge all the commits and let's keep future PRs smaller. But then you might not have access to the intermediate commits. The decision is up to you...

**Comment by archibate** [comment]
I personally agree with plan A, creating new PR by cherry-pick is so painful.

> I'll just squash and merge all the commits and let's keep future PRs smaller. But then you might not have access to the intermediate commits.

but you may not squash all commits but just squash some of these related each other, like add-range-for-support and fix-typo-in-thread-id. Also squash some minor fixes like no-test-opengl-in test_dynamic_for and opengl-raw_mod-fix.

**Comment by archibate** [comment]
I see we have too much features in this single PR: ext_arr, range_for, atomic_op, ... maybe we want to make them each a commit instead of each a PR?

**Comment by yuanming-hu** [comment]
> but you may not squash all commits but just squash some of these related each other, like add-range-for-support and fix-typo-in-thread-id. Also squash some minor fixes like no-test-opengl-in test_dynamic_for and opengl-raw_mod-fix to the commit in which they were implemented.

Then that's plan B - I'm good with that as well if you take some time and break it down. 

> I see we have too much features in this single PR: ext_arr, range_for, atomic_op, ... maybe we want to make them each a commit instead of each a PR?

If we do "each a PR" ultimately it will become each a commit after merging in the master branch anyway :-)

**Comment by archibate** [comment]
You mean, all PRs merged into master is to be squash into one commit?

**Comment by yuanming-hu** [comment]
> You mean, all PRs merged into master is to be squash into one commit?

Yeah, unless the commits might lead to indirect bugs somewhere else in the repo, or you really want to keep the intermediate commits. Also note that `Squash and merge` typically leads to fewer conflicts compared to `rebase and merge`, especially for large changesets.

**Comment by archibate** [comment]
I think I have understanded how PRs work in our repo: PRs in eyes of contributor = commits in eyes of taichi-dev. If we want to locate/blame something, first look at which function, eg. range_for-in-opengl, we search commits in taichi-dev's master, and found the corresponding PR is #594, and continue to search commits in that PR! We may futher more develop this pattern into:
taichi-dev = yuanming-hu + k-ye + taichi-opengl
taichi-opengl = archibate + glgeeker + helloworld
Where commits are in left, PRs are in right. We can have a clear structure of blaming system, may useful if taichi-dev keep grows, we may break down taichi into different fields like: syntax, llvm, opengl, cuda... each have their own branch, merge into master when wanted with ext PR.

File: taichi/perf.h (hunk 3 above)
Comment (archibate) [reply taichi/perf.h]: Sorry but I didn't realized there was an existing profiler..

**Comment by yuanming-hu** [comment]
> I think I have understanded how PRs work in our repo: PRs in eyes of contributor = commits in eyes of taichi-dev.

Right, I guess this is also the case for many other projects on github. This also allows us to use the more formal PR title as the commit message, leading to a clearer development history.



> We may futher more develop this pattern into:
> taichi-dev = yuanming-hu + k-ye + taichi-opengl
> taichi-opengl = archibate + glgeeker + helloworld
> Where commits are in left, PRs are in right. We can have a clear structure of blaming system, may useful if taichi-dev keep grows.
> Yeah, creating taichi-opengl the orgnization/user may be an overkill :) May just break down taichi into different fields like: syntax, llvm, opengl, cuda... each have their own branch, merge into master when wanted with ext PR, when I want to make gl works for example, I just make small PRs for different features into opengl. And then merge the whole opengl with master in another big PR.

I'd rather not make things complicated. Let's simply merge everything into master for now.

**Comment by archibate** [comment]
TODO: do good test on this branch with master.

**Comment by archibate** [comment]
can't merge anymore. almost forget everything did in gl.

**Comment by archibate** [comment]
all test passing..

**Comment by yuanming-hu** [comment]
> all test passing..

Great!! Could you merge your latest commit in this PR with the current master branch?

**Comment by archibate** [comment]
Just take all from g2 and drop all from master, hope this works..

**Comment by archibate** [comment]
Didn't work.

**Comment by archibate** [comment]
Giving up now.

**Comment by archibate** [comment]
```
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:933:21: error: C++ requires a type specifier for all declarations
  KernelGen codegen(kernel_, kernel_name_, struct_compiled_, global_tmps_buffer_size_);
                    ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:933:30: error: C++ requires a type specifier for all declarations
  KernelGen codegen(kernel_, kernel_name_, struct_compiled_, global_tmps_buffer_size_);
                             ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:933:44: error: C++ requires a type specifier for all declarations
  KernelGen codegen(kernel_, kernel_name_, struct_compiled_, global_tmps_buffer_size_);
                                           ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:933:62: error: unknown type name 'global_tmps_buffer_size_'
  KernelGen codegen(kernel_, kernel_name_, struct_compiled_, global_tmps_buffer_size_);
                                                             ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:934:3: error: unknown type name 'codegen'
  codegen.run(*prog_->snode_root);
  ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:934:10: error: expected member name or ';' after declaration specifiers
  codegen.run(*prog_->snode_root);
  ~~~~~~~^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:935:3: error: 'auto' not allowed in non-static struct member
  auto compiled = codegen.get_compiled_program();
  ^~~~
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:936:3: error: expected member name or ';' after declaration specifiers
  return [compiled](Context &ctx) {
  ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:942:2: error: expected ';' after struct
}
 ^
/home/bate/Develop/taichi/taichi/codegen/codegen_opengl.cpp:149:8: warning: private field 'has_rand_' is not used [-Wunused-private-field]
  bool has_rand_{false};
```
Now I'm completely lose control with these codes... Have no idea which stage is this on... The order is... all broken... what's more the good news is there still some unmerged commits in the old unstaged PR.. from the lesson learnt from GL: DON'T BREAK DOWN PRS ANYMORE! Only open-review-and-merge-in-time this method start from beginning can have benefits instead of wasting time :(

**Comment by yuanming-hu** [comment]
How about merging 664e1bd13ce9ea3b3e1eb0741ef12ed4cafe2674 with master? It seems to me that if you do that then the only file with conflict is `codegen_opengl.cpp`. Does fixing the merge conflicts, in that case, solve your compilation issues?

**Comment by archibate** [comment]
> Does fixing the merge conflicts, in that case, solve your compilation issues?
Didn't. The problem is, I cherry-picked commits in different timeline. And when merge, the timeline is broken, can't find a good order to make them all happy.

**Comment by archibate** [comment]
> How about merging 664e1bd with master? It seems to me that if you do that then the only file with conflict is codegen_opengl.cpp.

True, only 1 file conflict, but 18 conflicts with 100+ lines.

**Comment by yuanming-hu** [comment]
> > How about merging [664e1bd](https://github.com/taichi-dev/taichi/commit/664e1bd13ce9ea3b3e1eb0741ef12ed4cafe2674) with master? It seems to me that if you do that then the only file with conflict is codegen_opengl.cpp.
> 
> True, only 1 file conflict, but 18 conflicts with 100+ lines.

I see. Fixing that file seems the only way we can go now, unfortunately. Sorry about the trouble. Let's keep future PRs smaller to avoid related issues.

**Comment by archibate** [comment]
> > > How about merging [664e1bd](https://github.com/taichi-dev/taichi/commit/664e1bd13ce9ea3b3e1eb0741ef12ed4cafe2674) with master? It seems to me that if you do that then the only file with conflict is codegen_opengl.cpp.
> > 
> > 
> > True, only 1 file conflict, but 18 conflicts with 100+ lines.
> 
> I see. Fixing that file seems the only way we can go now, unfortunately. Sorry about the trouble. Let's keep future PRs smaller to avoid related issues.

Yeah you don't want to write any code about GL until all stages are merged.
No problem if we merge archibate:g1, archibate:g2 into archibate:opengl, I have write access to that branch, staging safely without requesting yuanming-hu, after all done do some final check and merge opengl into master, while small PRs are still trackable in/from archibate:opengl.

**Comment by archibate** [comment]
Got many F in test now.. Why not go back and use plan B..

**Comment by archibate** [comment]
Fixed, seems due to approx typo fix, need 1e-3.

**Comment by yuanming-hu** [comment]
I understand things are getting slightly more troublesome than expected.

> No problem if we merge archibate:g1, archibate:g2 into archibate:opengl, I have write access to that branch, staging safely without requesting yuanming-hu, after all done do some final check and merge opengl into master, while small PRs are still trackable in/from archibate:opengl. I heard lots of great github repos are using this way to modularize development.

I feel like merging a huge `archibate:opengl` branch as a PR, in the end, will still lead to some merge conflicts.

Having smaller PRs into master gradually will 
 - Allow reviewers to provide early feedback
 - Since PR is small,  
   - Review can be done quickly
   - Reviewers' job is easier and bugs can be more easily spotted
   - Potential merge conflicts are minimized (since small PR merge is quick, master has not changed too much by the time the PR is approaved)

What do you think?

**Comment by archibate** [comment]
I see, fault was I didn't even wait for your review/merge and keep pushing into opengl, causing the PR huge and breakless, i.e. **it should be small PRs at the first place, rather than breaking down an existing huge PR**.

**Comment by yuanming-hu** [comment]
Yeah, I should have let you know the PR is getting large, at an earlier time...
I think we both learned something from this :-) Thanks for your hard work and patience.

I'll wait for CI to pass - meanwhile, do we still need `taichi/perf.h`? We already have the `TI_PROFILER` macro.

**Comment by archibate** [comment]
I T   P A S S E D !
Merge when you think it's time for it.
BTW, can you make `ti test` only test for `ti.opengl` not for `ti.x64`? Can save lots of time for alter backend developers.

**Comment by yuanming-hu** [comment]
> I T P A S S E D !
> Merge when you think it's time for it.

Awesome!!

> BTW, can you make `ti test` only test for `ti.opengl` not for `ti.x64`? Can save lots of time for alter backend developers.

Great idea!! What do you think should be a good CLI for it? Free free to open up a stand-alone issue for this. (This PR now has 50+ conversations, LOL)

Btw, currently `python/taichi/main.py` for the `ti` command is slightly messy. We should consider using `argparse` if that makes sense.

**Comment by yuanming-hu** [comment]
Finally, we can merge this in! Thank you so much for your contribution. It's absolutely a giant step :-)

Going to sleep now. It's almost 3 am here. See you tmr.

**Pull request merged**